File: java/structures/DisjointSets.java
package structures;

// https://en.wikipedia.org/wiki/Disjoint-set_data_structure with path compression heuristic
// Complexity of operations is O(log(n)) on average
public class DisjointSets {

    public static int[] createSets(int size) {
        int[] p = new int[size];
        for (int i = 0; i < size; i++)
            p[i] = i;
        return p;
    }

    public static int root(int[] p, int x) {
        return x == p[x] ? x : (p[x] = root(p, p[x]));
    }

    public static boolean unite(int[] p, int a, int b) {
        a = root(p, a);
        b = root(p, b);
        if (a != b) {
            p[a] = b;
            return true;
        }
        return false;
    }

    // Usage example
    public static void main(String[] args) {
        int[] p = createSets(10);
        System.out.println(false == (root(p, 0) == root(p, 9)));
        unite(p, 0, 9);
        System.out.println(true == (root(p, 0) == root(p, 9)));
    }
}

File: golang/structures/disjoint_sets.go
package structures

// DisjointSets - Union-Find with path compression and union by rank
// https://en.wikipedia.org/wiki/Disjoint-set_data_structure
type DisjointSets struct {
	parent []int
	rank   []int
}

func NewDisjointSets(size int) *DisjointSets {
	ds := &DisjointSets{
		parent: make([]int, size),
		rank:   make([]int, size),
	}
	for i := range ds.parent {
		ds.parent[i] = i
	}
	return ds
}

func (ds *DisjointSets) Find(x int) int {
	if ds.parent[x] != x {
		ds.parent[x] = ds.Find(ds.parent[x])
	}
	return ds.parent[x]
}

func (ds *DisjointSets) Unite(a, b int) bool {
	a = ds.Find(a)
	b = ds.Find(b)
	if a == b {
		return false
	}
	if ds.rank[a] < ds.rank[b] {
		a, b = b, a
	}
	ds.parent[b] = a
	if ds.rank[a] == ds.rank[b] {
		ds.rank[a]++
	}
	return true
}
